File: JeMPI_Apps/JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph/CustomDgraphInteraction.java
package org.jembi.jempi.libmpi.dgraph;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.jembi.jempi.shared.models.InteractionWithScore;
import org.jembi.jempi.shared.models.CustomUniqueInteractionData;
import org.jembi.jempi.shared.models.CustomDemographicData;
import org.jembi.jempi.shared.models.Interaction;

@JsonInclude(JsonInclude.Include.NON_NULL)
record CustomDgraphInteraction(
      @JsonProperty("uid") String interactionId,
      @JsonProperty("Interaction.source_id") DgraphSourceId sourceId,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_AUX_DATE_CREATED) java.time.LocalDateTime auxDateCreated,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_AUX_ID) String auxId,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_AUX_CLINICAL_DATA) String auxClinicalData,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_GIVEN_NAME) String givenName,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_FAMILY_NAME) String familyName,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_GENDER) String gender,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_DOB) String dob,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_CITY) String city,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_PHONE_NUMBER) String phoneNumber,
      @JsonProperty(CustomDgraphConstants.PREDICATE_INTERACTION_NATIONAL_ID) String nationalId,
      @JsonProperty("GoldenRecord.interactions|score") Float score) {

   CustomDgraphInteraction(
         final Interaction interaction,
         final Float score) {
      this(interaction.interactionId(),
           new DgraphSourceId(interaction.sourceId()),
           interaction.uniqueInteractionData().auxDateCreated(),
           interaction.uniqueInteractionData().auxId(),
           interaction.uniqueInteractionData().auxClinicalData(),
           interaction.demographicData().givenName,
           interaction.demographicData().familyName,
           interaction.demographicData().gender,
           interaction.demographicData().dob,
           interaction.demographicData().city,
           interaction.demographicData().phoneNumber,
           interaction.demographicData().nationalId,
           score);
   }

   Interaction toInteraction() {
      return new Interaction(this.interactionId(),
                             this.sourceId() != null
                                   ? this.sourceId().toSourceId()
                                   : null,
                             new CustomUniqueInteractionData(this.auxDateCreated,
                                                               this.auxId,
                                                               this.auxClinicalData),
                             new CustomDemographicData(this.givenName,
                                                       this.familyName,
                                                       this.gender,
                                                       this.dob,
                                                       this.city,
                                                       this.phoneNumber,
                                                       this.nationalId));
   }

   InteractionWithScore toInteractionWithScore() {
      return new InteractionWithScore(toInteraction(), this.score());
   }

}

File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomDgraphInteraction.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomDgraphInteraction {

  private val classLocation =
    "../JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph"
  private val customClassName = "CustomDgraphInteraction"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(config: Config): Unit =
    val classFile: String =
      classLocation + File.separator + customClassName + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)

    writer.println(s"""package $packageText;
         |
         |import com.fasterxml.jackson.annotation.JsonInclude;
         |import com.fasterxml.jackson.annotation.JsonProperty;
         |import org.jembi.jempi.shared.config.DGraphConfig;
         |
         |@JsonInclude(JsonInclude.Include.NON_NULL)
         |record $customClassName(
         |      @JsonProperty("uid") String interactionId,
         |      @JsonProperty("Interaction.source_id") DgraphSourceId sourceId,
         |${interactionFields()}
         |      @JsonProperty("GoldenRecord.interactions|score") Float score) {
         |
         |}
         |""".stripMargin)
    writer.flush()
    writer.close()

    def interactionFields(): String =

      def mapField(
          predicate: String,
          fieldName: String,
          fieldType: String
      ): String =
        s"""${" " * 6}@JsonProperty($predicate) ${Utils
            .javaType(fieldType)} ${Utils.snakeCaseToCamelCase(fieldName)},"""

      val f1 =
        if (config.uniqueInteractionFields.isEmpty) ""
        else
          config.uniqueInteractionFields.get
            .map(f =>
              mapField(
                s"""DGraphConfig.PREDICATE_INTERACTION_${f.fieldName.toUpperCase}""",
                f.fieldName,
                f.fieldType
              )
            )
            .mkString(sys.props("line.separator")) + sys.props("line.separator")

      val f2 = config.demographicFields.zipWithIndex
        .map((f, i) =>
          mapField(
            s"""${"\"Interaction.demographic_field_%02d\"".format(i)}""",
            f.fieldName,
            f.fieldType
          )
        )
        .mkString(sys.props("line.separator"))

      f1 + f2
    end interactionFields

    def interactionConstructorArguments(): String =
      val f1 =
        if (config.uniqueInteractionFields.isEmpty) ""
        else
          config.uniqueInteractionFields.get
            .map(f =>
              s"""${" " * 11}interaction.uniqueInteractionData().${Utils
                  .snakeCaseToCamelCase(f.fieldName)}(),"""
            )
            .mkString(sys.props("line.separator")) + sys.props("line.separator")

      val f2 = config.demographicFields.zipWithIndex
        .map((_, idx) =>
          s"""${" " * 11}interaction.demographicData().fields.get($idx).value(),"""
        )
        .mkString(sys.props("line.separator"))

      f1 + f2
    end interactionConstructorArguments

    def uniqueArguments(): String =
      if (config.uniqueInteractionFields.isEmpty)
        ""
      else
        config.uniqueInteractionFields.get
          .map(f =>
            s"""${" " * 63}this.${Utils.snakeCaseToCamelCase(f.fieldName)},"""
          )
          .mkString(sys.props("line.separator"))
          .trim
          .dropRight(1)
      end if
    end uniqueArguments

    def demographicArguments(): String =
      config.demographicFields
        .map(f =>
          s"""${" " * 79}this.${Utils.snakeCaseToCamelCase(f.fieldName)},"""
        )
        .mkString(sys.props("line.separator"))
        .trim
        .dropRight(1)
    end demographicArguments

  end generate

}
